proposal/public-compiler-api.md

```diff
@@ -0,0 +1,172 @@
+module system are also impossible. Consider this case using webpack:
+
+**App.js**
+
+```js
+import 'styles.scss'
+```
+
+**styles.scss**;
+
+```scss
+@use 'bootstrap' with (
+  $btn-color: red
+)
+
+html {
+  font-size: 10px;
+}
+```
+
+Further down in the app, a component may wish to reference the previously configured
+library
+
+**button.scss**
+
+```scss
+@use 'bootstrap/buttons' as buttons
+
+.my-btn {
+  color: buttons.$btn-color;
+}
+```
+
+And the js file using webpack's `import` syntax for styles:
+
+**Button.js**
+
+```js
+import 'button.scss'
+
+// some web, vue, react, etc component
+```
+
+If the user is used module caching behavior like in Node.js, they would be suprised
+that `$btn-color` is its default not `red`.
+
+[css-modules]: https://github.com/css-modules/css-modules
+
+## Summary
+
+Sass implementations should provide a constructable Compiler class whose instance exposes `render` and `renderSync` methods for compiling Scss. Method calls providing a `file` have their results (as well as the results of and dependencies) cached by absolute filename on the compiler instance. Further calls to `render(Sync)` will return the cached value if not invalidated. Calls providing `data` would only have dependencies cached. Calls providing **both** `data` and `file`
+would process `data` using the resolved absolute path of `file` as it's cache key.
+
+The compiler instance would also expose an `invalidate` method which marks a file
+and all of it's dependents as needing recompilation.
+
+## Design choices
+
+> This section is non-normative.
+
+It's important that the compiler API allow delegating cache control to the consumer.
+Bundlers like webpack, or Parcel provide their own mechanisms for caching, watching, invalidating and topographically rebuilding entrypoints based on file changes. The Sass API
+should allow integration into those mechanisms.
+
+This does not preclude the possibility of Sass implementations providing their own
+watching mechanisms. This feature could be built on top of the proposed API
+
+## API
+
+```ts
+interface CompilerClass {
+  new (): Compiler
+}
+
+interface Compiler {
+  render(options: { data?: string, file?: string, ... }): Promise<string>
+
+  renderSync(options: { data?: string, file?: string, ... }): string
+
+  invalidate(file: string): void
+}
+
+```
+
+## Semantics
+
+- When `compiler.render(Sync)` is called
+
+  - Let `import-cache` be a `Map<string, string>` on `compiler`
+
+  - Let `module-cache` be a `Map<string, module>` on `compiler`
+
+  - If options provides only `file`
+
+    - Let `data` be the contents of `file`
+
+  - If `file` is set
+
+    - Let `filepath` be the absolute path of `file`
+
+  - If `filepath` and `import-cache` has key `filepath`
+
+    - Return string value for key `filepath`
+
+  - Overwise compile `data`
+
+    - For each `@import` rule
+
+      - Let `filepath` be the absolute path of the import
+
+      - If `import-cache` has key `filepath`
```

Comment:
Yes this is definitely hand wavey. I was thinking tho at some point urls are resolved to actual file paths and that could be used as the cache key. I had not considered the custom importer case :/

Reply:
As far as the core Sass compiler is concerned, URLs are actually *never* resolved to file paths. That's entirely the domain of importers. It just happens to be the case that the default importer is one that handles `file:` URLs.

I recommend taking a look at [Dart Sass's `Importer` API](https://pub.dev/documentation/sass/latest/sass/Importer-class.html) for an idea of how importers work under the covers. The short answer is that all files have a single canonical URL which has a one-to-one relationship with file context, and that importers convert URLs as written in stylesheets into fully canonical URLs.

Reply:
That makes sense, so `canonicalize` would return the correct key it seems like? And if i'm reading the import cache code correctly `Importer.canonicalize` will return an empty value if it can't handle the import so it should is it enough to have the logic be something like "canonicalize the url using the first importer that returns a valid value"

Reply:
Yeah, that would work. You could even gloss over the entire importer infrastructure if you wanted and just refer to the map keys as "canonical URLs".